File: ch2_labs/lab4_C++/FigureLibrary/Figure.cpp
#include "Figure.h"
#include <cmath>

Point::Point(double xVal, double yVal) : x(xVal), y(yVal) {}

double Point::distanceTo(const Point& other) const {
    double dx = x - other.x;
    double dy = y - other.y;
    return std::sqrt(dx * dx + dy * dy);
}

Figure::Figure(const std::vector<Point>& points) : vertices(points) {}

double Figure::getSideLength(int index) const {
    int nextIndex = (index + 1) % vertices.size();
    return vertices[index].distanceTo(vertices[nextIndex]);
}

std::vector<Point> Figure::getVertices() const {
    return vertices;
}

File: ch2_labs/lab4_C#/FigureLibrary/Figure.cs
namespace FigureLibrary
{
    public class Point
    {
        public double X { get; }
        public double Y { get; }

        public Point(double x, double y)
        {
            X = x;
            Y = y;
        }

        public double DistanceTo(Point other)
        {
            double dx = X - other.X;
            double dy = Y - other.Y;
            return Math.Sqrt(dx * dx + dy * dy);
        }
    }

    public class Figure
    {
        protected Point[] vertices;

        public Figure(Point[] points)
        {
            vertices = points;
        }

        public double SideLength(int index)
        {
            int nextIndex = (index + 1) % vertices.Length;
            return vertices[index].DistanceTo(vertices[nextIndex]);
        }

        public Point[] GetVertices()
        {
            return vertices;
        }
    }
}
